### PR TITLE
WIP: Conv seq2seq implementation

### PR DESCRIPTION
**Pull request by VahidooX**

### CODE REVIEW

File: open_seq2seq/parts/convs2s/utils.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from __future__ import absolute_import
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/utils.py]: Why is this file necessary? We have all these functions under parts/transformer/utils.py 
You should just re-use "get_padding" and "get_padding_bias" from there.

File: open_seq2seq/parts/convs2s/ffn_wn_layer.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+class FeedFowardNetworkNormalized(tf.layers.Layer):
+  """Fully connected feedforward network with weight normalization"""
+  """Inspired from https://github.com/tobyyouup/conv_seq2seq"""
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/ffn_wn_layer.py]: "Inspired by ...." should go to the top of the file

File: open_seq2seq/parts/convs2s/ffn_wn_layer.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  def __init__(self, in_dim, out_dim, dropout, var_scope_name):
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/ffn_wn_layer.py]: can you please add documentation here.

File: open_seq2seq/parts/convs2s/embedding_layer.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# Copyright 2018 MLBenchmark Group. All Rights Reserved.
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/embedding_layer.py]: This file is not needed. Use ``EmbeddingSharedWeights`` class from parts/transformer/embedding_layer.py

File: open_seq2seq/parts/convs2s/conv_wn_layer.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    return output
+
+  def gated_linear_units(self, inputs):
+    input_shape = inputs.get_shape().as_list()
+    assert len(input_shape) == 3
+    input_pass = inputs[:, :, 0:int(input_shape[2] / 2)]
+    input_gate = inputs[:, :, int(input_shape[2] / 2):]
+    input_gate = tf.sigmoid(input_gate)
+    return tf.multiply(input_pass, input_gate)
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/conv_wn_layer.py]: end all files with newline

File: open_seq2seq/parts/convs2s/attention_wn_layer.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from open_seq2seq.parts.convs2s.ffn_wn_layer import FeedFowardNetworkNormalized
+
+class AttentionLayerNormalized(tf.layers.Layer):
+  """Attention layer for convs2s with weight normalization"""
+  """ Inspired from https://github.com/tobyyouup/conv_seq2seq"""
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/attention_wn_layer.py]: "Inspired by ..." should go to the top of the file

File: open_seq2seq/parts/convs2s/attention_wn_layer.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  def __init__(self, in_dim, embed_size, layer_id, add_res):
```
Comment (okuchaiev) [inline open_seq2seq/parts/convs2s/attention_wn_layer.py]: Please add documentation here

File: open_seq2seq/encoders/convs2s_encoder.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    self._src_vocab_size = self.params['src_vocab_size']
+    self._src_emb_size = self.params['src_emb_size']
+    self.layers = []
+    self._mode = mode
+
+
+  def _encode(self, input_dict):
+    """
+    Encodes data into representation
```
Comment (okuchaiev) [inline open_seq2seq/encoders/convs2s_encoder.py]: This documentation format looks wrong. It should have "Args:" and (optionally) "Returns" sections. 
See ``open_seq2seq/encoders/encoder.py:151`` for example

File: example_configs/text2text/en-de/en-de-convs2s.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
+from __future__ import absolute_import, division, print_function
```
Comment (Kipok) [inline example_configs/text2text/en-de/en-de-convs2s.py]: Can you add "from \_\_future\_\_ import unicode_literals" to all files? And check your python2 fix again, because it might affect that code.

File: example_configs/text2text/en-de/en-de-convs2s.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+  "batch_size_per_gpu": batch_size,
+  "save_summaries_steps": 50,
+  "print_loss_steps": 50,
+  "print_samples_steps": 50,
+  "eval_steps": 4000,
+  "save_checkpoint_steps": 1000,
+  "logdir": "ReadData-CC",
+
+
+  # "optimizer": tf.contrib.opt.LazyAdamOptimizer,
```
Comment (Kipok) [inline example_configs/text2text/en-de/en-de-convs2s.py]: Let's remove all unused comments before we merge it

File: example_configs/text2text/nmt-reversal-CC.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
+from __future__ import absolute_import, division, print_function
```
Comment (Kipok) [inline example_configs/text2text/nmt-reversal-CC.py]: I suggest to put all this "toy" configs (yours and RR, RT, TT) into the "toy-nmt-reversal" subfolder (or a similar name), because there are already too many of this configs and they are here only for the sake of sanity check, not doing anything very useful.

File: example_configs/text2text/en-de/en-de-convs2s.py (hunk 10 above)
Comment (Kipok) [reply example_configs/text2text/en-de/en-de-convs2s.py]: From all the other files as well

File: open_seq2seq/decoders/convs2s_decoder.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    self.embedding_softmax_layer = None
+    self.position_embedding_layer = None
+    self.layers = []
+    self._tgt_vocab_size = self.params['tgt_vocab_size']
+    self._tgt_emb_size = self.params['tgt_emb_size']
+    self._mode = mode
+    self._pad_sym = self.params.get('PAD_SYMBOL', 0)
+    self._pad2eight = params.get('pad_embeddings_2_eight', False)
+
+
```
Comment (Kipok) [inline open_seq2seq/decoders/convs2s_decoder.py]: Can you quickly check the codestyle to be PEP-8 complaint (except for the 2 spaces indents). E.g. here you should have 1 blank line instead of 2. And newlines at the end of each file.

Comment (Kipok) [reply open_seq2seq/decoders/convs2s_decoder.py]: And the max line width should be 90 chars everywhere

File: open_seq2seq/decoders/convs2s_decoder.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    encoder_outputs = input_dict['encoder_output']['outputs']
+    encoder_outputs_b = input_dict['encoder_output'].get('outputs_b', encoder_outputs)
+
+    inputs_attention_bias = input_dict['encoder_output'].get('inputs_attention_bias_cs2s', None)
+
+    with tf.name_scope("decode"):
+      # prepare decoder layers
+      if len(self.layers) == 0:
+        knum_list = self.params.get("conv_knum")
```
Comment (Kipok) [inline open_seq2seq/decoders/convs2s_decoder.py]: .get will append None as the default if default explicitly not specified. Did you mean to have self.params.get("conv_knum", []) here? Or not use .get and move this parameter to required_params?

File: open_seq2seq/decoders/convs2s_decoder.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+    # Get the top sequence for each batch element
+    top_decoded_ids = decoded_ids[:, 0, :]
+    top_scores = scores[:, 0]
+
+    # this isn't particularly efficient
+    logits = self.decode_pass(top_decoded_ids, encoder_outputs, encoder_outputs_b, inputs_attention_bias)
+
+    return {"logits": logits,
+            "samples": [top_decoded_ids],
```
Comment (Kipok) [inline open_seq2seq/decoders/convs2s_decoder.py]: "samples" has been renamed to "outputs". Can you also merge your branch with current dev0.4 before merging them back to avoid conflicts?

File: open_seq2seq/encoders/convs2s_encoder.py [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+      'max_input_length': int,
+      'PAD_SYMBOL': int,
+
+      'mask_paddings': bool
+    })
+
+  def __init__(self, params, model, name="convs2s_encoder_with_emb", mode='train'):
+    """
+    Initializes convolutional encoder with embeddings
```
Comment (Kipok) [inline open_seq2seq/encoders/convs2s_encoder.py]: Incorrect docs style, use Args, Returns

Comment (Kipok) [reply open_seq2seq/encoders/convs2s_encoder.py]: Same for the functions below

File: open_seq2seq/parts/convs2s/attention_wn_layer.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  def call(self, input, target_embed, encoder_output_a, encoder_output_b, input_attention_bias):
+    """Calculates the attention vectors.
+
+    Args:
+      input: A float32 tensor with shape [batch_size, length, in_dim]
+      target_embed: A float32 tensor with shape [batch_size, length, in_dim] containing the target embeddings
+      encoder_output_a:
+      encoder_output_b:
+      input_attention_bias:
+    Returns:
```
Comment (Kipok) [inline open_seq2seq/parts/convs2s/attention_wn_layer.py]: Need to have explicit blank line before Returns to render it correctly in the docs

File: example_configs/text2text/en-de/transformer-big.py [hunk 17]
```diff
@@ -18,7 +18,7 @@
 d_model = 512
 num_layers = 6
 
-data_root = "[REPLACE THIS TO THE PATH WITH YOUR WMT DATA]"
+data_root = "./wmt16_en_dt/"
```
Comment (Kipok) [inline example_configs/text2text/en-de/transformer-big.py]: Just one last thing: can you add a comment here to indicate that users should replace this path with path to their data? Or change it back to the placeholder with that message, but I would prefer having it as a comment
